File: ckl-java/checkerlang/src/ch/checkerlang/AsList.java
package ch.checkerlang;

import ch.checkerlang.values.Value;
import ch.checkerlang.values.ValueList;
import ch.checkerlang.values.ValueString;

import java.util.Map;

public class AsList {

    public static ValueList from(Value value, String what)  {
        if (value.isString()) {
            String s = value.asString().getValue();
            ValueList list = new ValueList();
            for (int i = 0; i < s.length(); i++) {
                list.addItem(new ValueString(s.substring(i, i + 1)));
            }
            return list;
        }

        if (value.isMap() && what.equals("keys")) {
            ValueList list = new ValueList();
            for (Value member : value.asMap().getValue().keySet()) {
                list.addItem(member);
            }
            return list;
        }

        if (value.isMap() && what.equals("entries")) {
            ValueList list = new ValueList();
            for (Map.Entry<Value, Value> entry : value.asMap().getValue().entrySet()) {
                ValueList element = new ValueList();
                element.addItem(entry.getKey());
                element.addItem(entry.getValue());
                list.addItem(element);
            }
            return list;
        }

        if (value.isObject() && what.equals("keys")) {
            ValueList list = new ValueList();
            for (String member : value.asObject().value.keySet()) {
                list.addItem(new ValueString(member));
            }
            return list;
        }

        if (value.isObject() && what.equals("values")) {
            ValueList list = new ValueList();
            for (Value member : value.asObject().value.values()) {
                list.addItem(member);
            }
            return list;
        }

        if (value.isObject() && what.equals("entries")) {
            ValueList list = new ValueList();
            for (Map.Entry<String, Value> entry : value.asObject().value.entrySet()) {
                ValueList element = new ValueList();
                element.addItem(new ValueString(entry.getKey()));
                element.addItem(entry.getValue());
                list.addItem(element);
            }
            return list;
        }

        return value.asList();
    }
}

File: ckl-dotnet/ckl-dotnet/AsList.cs
namespace CheckerLang
{
    public class AsList
    {
        public static ValueList From(Value value, string what)
        {
            if (value.IsString())
            {
                var s = value.AsString().GetValue();
                var list = new ValueList();
                for (int i = 0; i < s.Length; i++)
                {
                    list.AddItem(new ValueString(s.Substring(i, 1)));
                }

                return list;
            }

            if (value.IsMap() && what == "keys") {
                var list = new ValueList();
                foreach (var member in value.AsMap().GetValue().Keys) {
                    list.AddItem(member);
                }
                return list;
            }

            if (value.IsMap() && what == "entries") {
                var list = new ValueList();
                foreach (var entry in value.AsMap().GetValue()) {
                    var element = new ValueList();
                    element.AddItem(entry.Key);
                    element.AddItem(entry.Value);
                    list.AddItem(element);
                }
                return list;
            }

            if (value.IsObject() && what == "keys") {
                var list = new ValueList();
                foreach (var member in value.AsObject().value.Keys) {
                    list.AddItem(new ValueString(member));
                }
                return list;
            }

            if (value.IsObject() && what == "values") {
                var list = new ValueList();
                foreach (var member in value.AsObject().value.Values) {
                    list.AddItem(member);
                }
                return list;
            }

            if (value.IsObject() && what == "entries") {
                var list = new ValueList();
                foreach (var member in value.AsObject().value)
                {
                    var element = new ValueList();
                    element.AddItem(new ValueString(member.Key));
                    element.AddItem(member.Value);
                    list.AddItem(element);
                }
                return list;
            }

            if (value.IsObject())
            {
                var list = new ValueList();
                foreach (var member in value.AsObject().value.Keys) 
                {
                    list.AddItem(new ValueString(member));
                }
                return list;
            }

            return value.AsList();
        }
    }
}
